src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,913 @@
+2x≡0⇒x≡0 {x} 2x≡0 =  x≡0
+  where
+  2B*x≡0 =  trans (sym (2*≗2B* x)) 2x≡0
+
+  x≡0 :  x ≡ 0#
+  x≡0 with  x*y≡0⇒x≡0∨y≡0 2B 2B*x≡0
+  ... | inj₁ 2≡0 =  contradiction 2≡0 2suc≢0
+  ... | inj₂ eq  =  eq
+
+x≢0⇒2x≢0 :  ∀ {x} → x ≢ 0# → 2* x ≢ 0#
+x≢0⇒2x≢0 x≢0 =  x≢0 ∘ 2x≡0⇒x≡0
+
+------------------------------------------------------------------------------
+-- Commutativity and associativity for _*_, and distributivity of _*_ over +,
+-- are proved by using the isomorphisms to ℕ implemented and proved earlier.
+------------------------------------------------------------------------------
+
+*-comm : Commutative _*_
+*-comm a b = begin
+  a * b                    ≡⟨ sym (fromℕ∘toℕ (a * b)) ⟩
+  fromℕ (toℕ (a * b))      ≡⟨ cong fromℕ (toℕ*homo a b) ⟩
+  fromℕ (toℕ a *ₙ toℕ b)   ≡⟨ cong fromℕ (ℕp.*-comm (toℕ a) (toℕ b)) ⟩
+  fromℕ (toℕ b *ₙ toℕ a)   ≡⟨ cong fromℕ (sym (toℕ*homo b a)) ⟩
+  fromℕ (toℕ (b * a))      ≡⟨ fromℕ∘toℕ (b * a) ⟩
+  b * a                    ∎
+
+*-assoc :  Associative _*_
+*-assoc a b c = begin
+  (a * b) * c                 ≡⟨ sym (fromℕ∘toℕ ((a * b) * c)) ⟩
+  fromℕ (toℕ ((a * b) * c))   ≡⟨ cong fromℕ (toℕ*homo (a * b) c) ⟩
+  fromℕ (toℕ (a * b) *ₙ cN)   ≡⟨ cong (fromℕ ∘ (_*ₙ cN)) (toℕ*homo a b) ⟩
+  fromℕ ((aN *ₙ bN) *ₙ cN)    ≡⟨ cong fromℕ (ℕp.*-assoc aN bN cN) ⟩
+  fromℕ (aN *ₙ (bN *ₙ cN))    ≡⟨ cong (fromℕ ∘ (aN *ₙ_))
+                                                  (sym (toℕ*homo b c)) ⟩
+  fromℕ (aN *ₙ toℕ (b * c))   ≡⟨ cong fromℕ (sym (toℕ*homo a (b * c))) ⟩
+  fromℕ (toℕ (a * (b * c)))   ≡⟨ fromℕ∘toℕ (a * (b * c)) ⟩
+  (a * (b * c))               ∎
+  where
+  aN = toℕ a;   bN = toℕ b;   cN = toℕ c
+
+2*-*-assoc :  ∀ x y → (2* x) * y ≡ 2* (x * y)
+2*-*-assoc x y = begin
+  (2* x) * y     ≡⟨ cong (_* y) (2*≗2B* x) ⟩
+  (2B * x) * y   ≡⟨ *-assoc 2B x y ⟩
+  2B * (x * y)   ≡⟨ sym (2*≗2B* (x * y)) ⟩
+  2* (x * y)     ∎
+
+
+*-distribˡ-+ :  _DistributesOverˡ_ _*_ _+_
+*-distribˡ-+ a b c = begin
+  a * (b + c)                          ≡⟨ sym (fromℕ∘toℕ (a * (b + c))) ⟩
+  fromℕ (toℕ (a * (b + c)))            ≡⟨ cong fromℕ (toℕ*homo a (b + c)) ⟩
+  fromℕ (k *ₙ (toℕ (b + c)))           ≡⟨ cong (fromℕ ∘ (k *ₙ_))
+                                                         (toℕ+homo b c) ⟩
+  fromℕ (k *ₙ (m +ₙ n))                ≡⟨ cong fromℕ (ℕp.*-distribˡ-+ k m n) ⟩
+  fromℕ (k *ₙ m +ₙ k *ₙ n)             ≡⟨ cong fromℕ $
+                                          cong₂ _+ₙ_ (sym (toℕ*homo a b))
+                                                     (sym (toℕ*homo a c))
+                                        ⟩
+  fromℕ (toℕ (a * b) +ₙ toℕ (a * c))   ≡⟨ cong fromℕ $
+                                            sym (toℕ+homo (a * b) (a * c))
+                                        ⟩
+  fromℕ (toℕ (a * b + a * c))          ≡⟨ fromℕ∘toℕ (a * b + a * c) ⟩
+  a * b + a * c                        ∎
+  where
+  k = toℕ a;   m = toℕ b;   n = toℕ c
+
+
+*-distribʳ-+ :  _DistributesOverʳ_ _*_ _+_
+*-distribʳ-+ a b c = begin
```

Comment:
Done.
This simplifies the code. Thanks to @gallais.  

But may be it needs to be  ``\and``  rather than ``+`` in the names
``comm+cancelˡ⇒cancelʳ, comm+distrˡ⇒distrʳ`` -- ?
For I recall of the naming convention, and  the letter ``+`` refers in these names not to ``_+_`` but to \and.

Reply:
src/Data/Nat/Properties.agda
  suc (n + (m + m * n))  ≡⟨ cong suc (sym (+-assoc n m (m * n))) ⟩
  suc (n + m + m * n)    ≡⟨ cong (λ x → suc (x + m * n)) (+-comm n m) ⟩
  suc (m + n + m * n)    ≡⟨ cong suc (+-assoc m n (m * n)) ⟩
  suc (m + (n + m * n))  ≡⟨⟩

> This change doesn't seem to have been pushed.

Sorry, I do not find this notice in my ``comments*.lagda`` files (to where I gather them).
Probably I have lost it.
Can you, please, repeat it?
I guess this is about ``+-*-suc`` but do not guess of what is wrong there.

Reply:
src/Data/Nat/Properties.agda
m>1⇒m*n≢1 {m} {0}     _   m*0≡1  =  0≢1+n 0≡1
                                   where
                                   0≡1 = trans (sym (*-zeroʳ m)) m*0≡1

> Again this change doesn't seem to have been pushed.

Fixed.
I guess this is about indentation of ``where``.
(How strange: in the next sentence ``where`` is indented differently).